Qmail force-tls patch (roberto-netqmail-1.06_force-tls.patch-2011.10.06). See:
http://notes.sagredo.eu/sites/notes.sagredo.eu/files/qmail/patches/roberto-netqmail-1.06_force-tls.patch-2011.10.06
--- a/qmail-smtpd.c
+++ b/qmail-smtpd.c
@@ -362,7 +362,12 @@
   out("\r\n250-PIPELINING\r\n250-8BITMIME\r\n");
   out("250-SIZE "); out(size); out("\r\n");
 #ifdef CRAM_MD5
-  out("250 AUTH LOGIN PLAIN CRAM-MD5\r\n");
+/* forcetls patch */
+  char *cram_md5=env_get("CRAM_MD5");
+  if (!cram_md5 || (cram_md5 && strcmp(cram_md5, "1")!=0))
+    out("250 AUTH LOGIN PLAIN\r\n");
+  else out("250 AUTH LOGIN PLAIN CRAM-MD5\r\n");
+/* END forcetls patch */
 #else
   out("250 AUTH LOGIN PLAIN\r\n");
 #endif
@@ -674,6 +679,15 @@
 #ifdef CRAM_MD5
 int auth_cram()
 {
+/* forcetls patch */
+  char *cram_md5=env_get("CRAM_MD5");
+  if (!cram_md5 || (cram_md5 && strcmp(cram_md5, "1")!=0))
+  {
+    out("504 auth type unimplemented (#5.5.1) \r\n");
+    return;
+  }
+/* END forcetls patch */
+
   int i, r;
   char *s;
 
@@ -726,6 +740,22 @@
 void smtp_auth(arg)
 char *arg;
 {
+
+/* forcetls patch */
+#ifdef TLS
+  char *forcetls=env_get("FORCETLS");
+  if (!forcetls || (forcetls && strcmp(forcetls, "0")!=0))
+  {
+    if (!ssl)
+    {
+      out("538 auth not available without TLS (#5.3.3)\r\n");
+      flush();
+      die_read();
+    }
+  }
+#endif
+/* END forcetls patch */
+
   int i;
   char *cmd = arg;
 
